Route Caching

ascorbic opened the pull request:
<!--
**Warning**
**This template is reserved for approved proposals and Astro maintainers only.**
You are probably looking to [**start a discussion**](https://github.com/withastro/roadmap/discussions/new)!

Community members who would like to propose an idea or feature should begin
by creating a GitHub Discussion. See the repo [`README.md`](https://github.com/withastro/roadmap#readme) for more info.
-->

# Summary

A platform-agnostic route caching API for Astro SSR pages that enables declarative cache control using web standards. 

# Examples

## Basic route caching

```astro
---
// src/pages/products/[id].astro
import { getEntry } from 'astro:content';
const product = await getEntry('products', Astro.params.id);
Astro.cache.set({
  lastModified: product.updatedAt,
  maxAge: 300,  // Cache for 5 minutes
  swr: 3600,    // Stale-while-revalidate for 1 hour
  tags: ['products', `product:${product.id}`]
});
---
<h1>{product.data.name}</h1>
<p>{product.data.description}</p>
```

## Automatic dependency tracking

```ts
// src/pages/api/revalidate.ts
import { getLiveEntry } from "astro:content";

export const POST: APIRoute = async ({ cache, request }) => {
  const { id } = await request.json();
  const { entry } = await getLiveEntry("products", id);

  // Invalidate all pages that depend on this entry
  cache.invalidate(entry);

  return Response.json({ ok: true });
};
```


# Links
- [Rendered RFC](https://github.com/withastro/roadmap/blob/feat/route-caching/proposals/0056-route-caching.md)
- [Stage 2 Proposal](https://github.com/withastro/roadmap/issues/1140)
- [Stage 1 Proposal](https://github.com/withastro/roadmap/discussions/1131)

ascorbic committed:
Add route caching stage 3 RFC

ascorbic committed:
Update names

ascorbic committed:
Update Stage 3 PR link in route caching proposal

ascorbic committed:
Improve examples

Added cache configuration for routes in astro.config.ts and updated cache invalidation tag in webhook API.

florian-lefebvre commented on proposals/0056-route-caching.md:
```diff
@@ -0,0 +1,1149 @@
+  adapter: node(),
+  cache: {
+    // Override Node's default in-memory cache with Fastly
+    driver: "@astrojs/cache-fastly",
```
I think I'd prefer an API like fonts, where you call a function. That would make it easier to have custom drivers:
```js
driver: cacheFastly({
	serviceId: process.env.FASTLY_SERVICE_ID,
	token: process.env.FASTLY_TOKEN,
})
```

ascorbic replied:
Ah got it. Yeah, that could be a better pattern, as it makes it easier to type the options

florian-lefebvre replied:
LGTM!

ascorbic committed:
Updates based on review

florian-lefebvre commented:
I'm not sure that's something for Astro to provide, feels like maybe it should be achievable through the driver platform (eg. cloudflare)?

wildfiremedia commented:
My other thought is that if we could set a max memory limit, it might help prevent the server from starving or running out of memory when a malicious or heavy request comes in. Should it only cache those with HTTP respond code 200?

https://example.com?test=1
https://example.com?test=...
https://example.com?test=1000

ascorbic committed:
Update and clarify API

florian-lefebvre commented on proposals/0056-route-caching.md:
```diff
@@ -0,0 +1,1360 @@
+export default defineConfig({
+  adapter: node(),
+  cache: {
+    routes: {
```
This makes me think about <https://nitro.build/config#routerules> in nitro. So I'm wondering: is there any benefit to have this under `cache`? Instead that could be on the top level and we could also set `prerender: boolean` there.


So to recap, this looks good to me as is but maybe there's a way to future proof it and enable more cool things

ascorbic replied:
hmm, yeah. Good idea. routeRules would be a better name in that scneario too

ascorbic replied:
I wonder if we just directly copy the Nitro pattern, with "swr: true|number is shortcut for cache: { swr: true, maxAge: number }"

florian-lefebvre replied:
A recurring problem with supporting different shapes it doesn't work very nicely with `updateConfig()`. Otherwise no strong opinion

florian-lefebvre commented on proposals/0056-route-caching.md:
```diff
@@ -0,0 +1,1360 @@
+
+Caching is essential for performant server-rendered applications, but current solutions have significant limitations:
+
+**Platform-specific implementations**: Most caching solutions (like ISR) are tightly coupled to specific hosting platforms. ISR was created for Next.js on Vercel and while other platforms have implemented versions of it, they're often limited or second-class implementations. This creates vendor lock-in and inconsistent behavior across deployments.
```
What does that mean for <https://docs.astro.build/en/guides/integrations-guide/vercel/#isr>? Would we remove it in favor of `Astro.cache` using <https://vercel.com/changelog/tag-based-cache-invalidation-now-available-for-all-responses> under the hood?

ascorbic replied:
I think so, or leave it deprecated and make it conflice with the new caching

florian-lefebvre commented on proposals/0056-route-caching.md:
```diff
@@ -0,0 +1,1360 @@
+# Non-Goals
+
+- **Strong consistency guarantees**: Cache invalidation follows eventual consistency model
+- **Distributed caching for Node.js**: Node adapter uses in-memory cache, unsuitable for multi-instance deployments without external cache
```
I assume you mean "by default" but it can be configured

florian-lefebvre commented on proposals/0056-route-caching.md:
```diff
@@ -0,0 +1,1360 @@
+- **Distributed caching for Node.js**: Node adapter uses in-memory cache, unsuitable for multi-instance deployments without external cache
+- **Static/prerendered page caching**: This is for on-demand SSR routes only; static and prerendered routes are already cached by default and don't need route-level cache control
+- **Partial page caching**: This focuses on full-page route caching, not fragment or component-level caching
+- **Browser caching**: This focuses on CDN and server-side caching, not browser caching. Only `Last-Modified` and `ETag` headers are sent to browsers for conditional requests
```
Can users access the current state of the cache (eg. after several `cache.set()` calls) so they can use this data to add browser caching headers? So not only `Astro.cache.tags`

florian-lefebvre commented on proposals/0056-route-caching.md:
````diff
@@ -0,0 +1,1360 @@
+
+```ts
+interface CacheProvider {
+  name: string;
````
It's probably more useful to have the name in the `cache.provider` than the runtime implementation, so it can be used in logging etc

ascorbic committed:
Use Astro route syntax

ascorbic committed:
Updates: standardise terminology to "cache provider". Move from cache.routes to top-level routeRules

ascorbic commented:
@florian-lefebvre I have done an update that among other things moves the rules from `cache.routes` to `routeRules`

ascorbic committed:
update RFC to match implementation (provider naming, routeRules, DisabledAstroCache, node LRU details)

ascorbic committed:
docs: update RFC for memoryCache() in astro/config

Memory cache moved to core. Config function is now memoryCache()
from astro/config instead of cacheMemory() from @astrojs/node/cache.

ascorbic committed:
Update to route rule syntax

ascorbic committed:
Add memory provider cache key info

florian-lefebvre commented on proposals/0056-route-caching.md:
```diff
@@ -0,0 +1,1406 @@
+
+- Start Date: 2025-10-15
+- Reference Issues: https://github.com/withastro/roadmap/discussions/1131, https://github.com/withastro/roadmap/discussions/181
+- Implementation PR: <!-- leave empty -->
```
```suggestion
- Implementation PR: https://github.com/withastro/astro/pull/15579
```

ascorbic committed:
Update proposals/0056-route-caching.md

Co-authored-by: Florian Lefebvre <contact@florian-lefebvre.dev>
